fetch/mesh_data.go

```diff
@@ -83,18 +97,26 @@ func (f *Fetch) getHashes(
 	for i, hash := range hashes {
 		if err := options.limiter.Acquire(ctx, 1); err != nil {
 			pendingMetric.Add(float64(i - len(hashes)))
-			return fmt.Errorf("acquiring slot to get hash: %w", err)
+			err = fmt.Errorf("acquiring slot to get hash: %w", err)
+			for _, h := range hashes[i:] {
+				options.callback(h, err)
+			}
+			return err
```

Comment:
It's rather likely to be a small subset of ATX IDs e.g. not downloaded due to `hs/1` request throttling, and these requests will be retried later

Reply:
The point still stands - `getHashes` might be called with 5.0 Mio hashes at once - if the limit cannot be acquired because throttling is active then millions of callbacks are called.

As far as I can see this only affects this section of the code:

https://github.com/spacemeshos/go-spacemesh/blob/ad4924b998ca085bad90ebb9df2fcc16c5ab9c6c/sync2/atxs.go#L110-L129

This will ~print a debug log with the exact same error for every ATX and~ increment every element in `cs.state`. This could be handled much simpler (and arguably more efficiently) without requiring to keep track of the retries of every single hash:

Arguably this is out of the scope of this PR but this should be addressed. It makes no sense to register a hash for a peer then requesting that hash in a batch and let the fetcher again reconstruct from which peer to fetch that hash from. Error handling is also bad, because for every fetched hash an error has to be returned via callback or aggregated in a `&fetcher.BatchError{}`. Instead imo it would be much simpler to just have a (blocking) `getHash` method that fetches a hash from a given peer and returns an error if something went wrong. Then the caller can easily parallize requests and match errors to peers & hashes.

Internally the fetcher can still group requests into single batches, request those from individual peers and deserialise the batched result. This also makes it easier to figure out what when wrong if something did go wrong. Right now we have a lot of log errors of the kind `validation ran for unknown hash` and `hash missing from ongoing requests` because the fetcher fails to match requests of hashes to peers and/or callers with how it is structured at the moment.

Reply:
We should do a serious `fetcher` cleanup after we switch to `syncv2` at least for ATXs and malfeasance proofs. Right now it's probably a bit too early as we'll have to update v1 syncers that are on their way out.
Simplifying blob fetch logic that currently uses promises etc. should be one of the goals, other being removing non-streaming fetcher client and server code.

Reply:
OK, I summarized all my previous concerns with the fetcher in this issue as well as liked some existing issued in it to make sure they are not forgotten: https://github.com/spacemeshos/go-spacemesh/issues/6545